evm/auxiliary/user_account_contract/contract.py

```diff
@@ -0,0 +1,211 @@
+from evm.constants import (
+    SECPK1_N,
+    ENTRY_POINT,
+)
+from evm.utils.hexadecimal import (
+    decode_hex,
+)
+from evm.utils.numeric import (
+    big_endian_to_int,
+)
+from evm.utils.keccak import (
+    keccak,
+)
+
+
+# S <= N // 2
+MAX_ALLOWED_S = SECPK1_N // 2
+VALIDATION_CODE_GAS = 3500
+GAS_RESERVE = 4500  # amount of gas reserved for returning
+GAS_RESERVE_OFFSET = 200
+NONCE_GETTER_ID = big_endian_to_int(keccak(b"get_nonce()")[:4])
+VALIDATION_CODE_GETTER_ID = big_endian_to_int(keccak(b"get_validation_code()")[:4])
+ENTRY_POINT_INT = big_endian_to_int(ENTRY_POINT)
+
+SIGNATURE_VERIFICATION_ADDRESS = 1
+SIGNATURE_VERIFICATION_GAS = 3000
+
+# calldata locations
+CALLDATA_FUNCTION_ID = 0
+CALLDATA_SIGNATURE = 0
+(
+    CALLDATA_V,
+    CALLDATA_R,
+    CALLDATA_S,
+    CALLDATA_NONCE,
+    CALLDATA_GASPRICE,
+    CALLDATA_VALUE,
+    CALLDATA_MIN_BLOCK,
+    CALLDATA_MAX_BLOCK,
+    CALLDATA_DESTINATION,
+    CALLDATA_DATA,
+) = (CALLDATA_SIGNATURE + i * 32 for i in range(10))
+
+# memory locations
+MEMORY_CALLDATA = 32
+MEMORY_NONCE = MEMORY_CALLDATA + CALLDATA_NONCE
+MEMORY_DATA = MEMORY_CALLDATA + CALLDATA_DATA
+MEMORY_SIGHASH = 0
+MEMORY_SENDER = 0
+MEMORY_RETURNCODE = 0
+
+# storage locations
+STORAGE_NONCE = 0
+
+
+def generate_account_lll_code(validation_code_address):
+    return [
+        'seq',
+
+        # handle get_nonce if called
+        ['if',
+            ['eq',
+                ['calldataload', CALLDATA_FUNCTION_ID],
+                NONCE_GETTER_ID],
```

Comment:
Shouldn't NONCE_GETTER_ID be multiplied by 2^224? It's the first four bytes that are the function sig, so getting the first 32 bytes would return that with 28 zero bytes at the end.

BTW are there tests for all of this? If so, the tests should catch this.

Reply:
You're right, will fix this.

Tests are here: https://github.com/jannikluhn/py-evm/blob/07587ac50734c62a7d8862c072fbcaeedaa93c97/tests/auxiliary/user-account/test_contract.py

They cheat a bit by importing the function ids, that's why they didn't catch this. Will add a test comparing against a manually generated one (using the Vyper code).